### ISSUE TITLE
Build problems with Visual Studio.

### ISSUE THREAD

**Issue by slembcke** [slembcke/Chipmunk2D#126]
Received via email:

> After having all the paths and directories right:
> -i had to edit "chipmunk.c" at line 46 and put instead " const char *cpVersionString = ""; "
> -i compiled the lib, and i installed the C++ wrapper from jhasse, chipmunkpp
> -i got an error from cpCollision.obj, and i didn't understand for a while and then i discovered that cpRobust.h and cpRobust.c weren't put in the Visual Studio solution, i re-compiled and then got a weird error and it cames from the "cpRobust.c" at line 1 it was written " #import " instead of " #include "
> 
> so just a few tricks to get it compiled but, i wanted to share you these so you will maybe fix them
> 
> INFOS:
> -downloaded by github https://github.com/slembcke/Chipmunk2D/tree/07d74973eaa285cc7eee80e79ef9b9d545d2a9d1
> 
> -used the vs13 folder/solution
> 
> -used Visual Studio 2015

**Comment by Fighter19** [comment]
Also I found the following problems when trying to compile the 7.0.1 tag.
CMake tells MSVC to use CPP to compile.
This will fix the initializer at cpPolyline.c:564 but breaks the macro trick in chipmunk.c :
```
#define STR(s) #s
#define XSTR(s) STR(s)
```
Also there is no PThread in MSVC which needs someone redoing the code for MSVC located at cpHastySpace.c

(I found those by manually including the CMakeLists.txt in src/ not in the root, as I intended to build this library together with my project using CMake)

I'm not sure whether these files are needed (cpHastySpace.c) but they are still automatically included by the CMake script.

**Comment by slembcke** [comment]
Oof. This is a month old and I still haven't gotten to it. :-\ Sorry about that. I don't think I've booted Windows for months now...

cpHastySpace.c is optional. It has optimizations for ARM-NEON, and some threading code that doesn't make a lot of sense anymore. I wouldn't bother with it at all really.

**Comment by Alan-FGR** [comment]
> Also there is no PThread in MSVC which needs someone redoing the code for MSVC located at cpHastySpace.c

https://github.com/slembcke/Chipmunk2D/pull/136

> cpHastySpace.c is optional. It has optimizations for ARM-NEON, and some threading code that doesn't make a lot of sense anymore. I wouldn't bother with it at all really.

I myself think it does make sense even on desktops. If people don't care about performance they'll probably just stick with the Box2D implementation that ships with their engine/framework, or a port to their language of choice. I picked Chipmunk because of its performance and multi-threading and SIMD are contributing factors for that.

**Comment by Berserker66** [comment]
Pinging this issue, as it got relevant for pymunk, which used mingwpy to compile itself and chipmunk. mingwpy is however now abandoned and as such msvc may be needed for the windows build, or at least should be supported.

I mostly leave this here for reference and hope that someone looks into it.
This is the log when trying to build with Visual Studio 2017, from what I can tell, the MAKE_REF macro in chimpunk_ffi.h needs updating.

```
1>------ Build started: Project: chipmunk, Configuration: Debug x64 ------
1>cpSweep1D.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpSpatialIndex.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpSpaceStep.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>e:\chipmunk2d-master\src\cpspacestep.c(251): warning C4090: 'function': different 'const' qualifiers
1>cpSpaceQuery.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpSpaceHash.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpSpaceDebug.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpSpaceComponent.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>e:\chipmunk2d-master\src\cpspacecomponent.c(65): warning C4090: 'function': different 'const' qualifiers
1>cpSpace.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>e:\chipmunk2d-master\src\cpspace.c(410): warning C4307: '*': integral constant overflow
1>cpSlideJoint.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpSimpleMotor.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpShape.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpRotaryLimitJoint.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpRobust.c
1>cpRatchetJoint.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpPolyShape.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpPivotJoint.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpPinJoint.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpHashSet.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpGrooveJoint.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpGearJoint.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>Generating Code...
1>Compiling...
1>cpDampedSpring.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpDampedRotarySpring.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpConstraint.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpCollision.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpBody.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpBBTree.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpArray.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>cpArbiter.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>chipmunk.c
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_private.h(304): warning C4090: 'function': different 'const' qualifiers
1>e:\chipmunk2d-master\src\chipmunk.c(93): warning C4028: formal parameter 2 different from declaration
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(42): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(42): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpVect (__cdecl *)(const cpFloat,const cpFloat)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(43): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(43): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpBool (__cdecl *)(const cpVect,const cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(44): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(44): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpVect (__cdecl *)(const cpVect,const cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(45): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(45): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpVect (__cdecl *)(const cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(46): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(46): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpVect (__cdecl *)(const cpVect,const cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(47): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(47): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpVect (__cdecl *)(const cpVect,const cpFloat)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(48): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(48): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpFloat (__cdecl *)(const cpVect,const cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(49): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(49): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpFloat (__cdecl *)(const cpVect,const cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(50): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(50): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpVect (__cdecl *)(const cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(51): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(51): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpVect (__cdecl *)(const cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(52): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(52): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpVect (__cdecl *)(const cpVect,const cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(53): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(53): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpVect (__cdecl *)(const cpFloat)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(54): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(54): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpFloat (__cdecl *)(const cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(55): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(55): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpVect (__cdecl *)(const cpVect,const cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(56): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(56): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpVect (__cdecl *)(const cpVect,const cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(57): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(57): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpFloat (__cdecl *)(const cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(58): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(58): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpFloat (__cdecl *)(const cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(59): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(59): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpVect (__cdecl *)(const cpVect,const cpVect,const cpFloat)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(60): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(60): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpVect (__cdecl *)(const cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(61): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(61): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpVect (__cdecl *)(const cpVect,const cpFloat)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(62): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(62): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpVect (__cdecl *)(cpVect,cpVect,cpFloat)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(63): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(63): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpFloat (__cdecl *)(const cpVect,const cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(64): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(64): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpFloat (__cdecl *)(const cpVect,const cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(65): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(65): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpBool (__cdecl *)(const cpVect,const cpVect,const cpFloat)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(67): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(67): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpFloat (__cdecl *)(cpFloat,cpFloat)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(68): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(68): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpFloat (__cdecl *)(cpFloat,cpFloat)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(69): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(69): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpFloat (__cdecl *)(cpFloat)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(70): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(70): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpFloat (__cdecl *)(cpFloat,cpFloat,cpFloat)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(71): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(71): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpFloat (__cdecl *)(cpFloat,cpFloat,cpFloat)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(72): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(72): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpFloat (__cdecl *)(cpFloat,cpFloat,cpFloat)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(74): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(74): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpBB (__cdecl *)(const cpFloat,const cpFloat,const cpFloat,const cpFloat)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(75): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(75): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpBB (__cdecl *)(const cpVect,const cpFloat,const cpFloat)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(76): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(76): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpBB (__cdecl *)(const cpVect,const cpFloat)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(77): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(77): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpBool (__cdecl *)(const cpBB,const cpBB)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(78): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(78): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpBool (__cdecl *)(const cpBB,const cpBB)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(79): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(79): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpBool (__cdecl *)(const cpBB,const cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(80): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(80): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpBB (__cdecl *)(const cpBB,const cpBB)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(81): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(81): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpBB (__cdecl *)(const cpBB,const cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(82): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(82): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpVect (__cdecl *)(cpBB)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(83): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(83): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpFloat (__cdecl *)(cpBB)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(84): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(84): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpFloat (__cdecl *)(cpBB,cpBB)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(85): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(85): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpFloat (__cdecl *)(cpBB,cpVect,cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(86): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(86): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpBool (__cdecl *)(cpBB,cpVect,cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(87): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(87): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpVect (__cdecl *)(const cpBB,const cpVect)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(89): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(89): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'void (__cdecl *)(cpSpatialIndex *)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(90): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(90): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'int (__cdecl *)(cpSpatialIndex *)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(91): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(91): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'void (__cdecl *)(cpSpatialIndex *,cpSpatialIndexIteratorFunc,void *)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(92): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(92): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'cpBool (__cdecl *)(cpSpatialIndex *,void *,cpHashValue)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(93): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(93): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'void (__cdecl *)(cpSpatialIndex *,void *,cpHashValue)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(94): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(94): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'void (__cdecl *)(cpSpatialIndex *,void *,cpHashValue)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(95): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(95): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'void (__cdecl *)(cpSpatialIndex *)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(96): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(96): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'void (__cdecl *)(cpSpatialIndex *,void *,cpHashValue)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(97): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(97): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'void (__cdecl *)(cpSpatialIndex *,void *,cpVect,cpVect,cpFloat,cpSpatialIndexSegmentQueryFunc,void *)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(98): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(98): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'void (__cdecl *)(cpSpatialIndex *,void *,cpBB,cpSpatialIndexQueryFunc,void *)'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(99): error C2143: syntax error: missing '{' before '*'
1>e:\chipmunk2d-master\include\chipmunk\chipmunk_ffi.h(99): warning C4047: 'initializing': 'int *' differs in levels of indirection from 'void (__cdecl *)(cpSpatialIndex *,cpSpatialIndexQueryFunc,void *)'
1>Generating Code...
1>Done building project "chipmunk.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========

```

**Comment by Berserker66** [comment]
I got it to build by changing line 30 in chipmunk_ffi.h to :

`  #define MAKE_REF(name) CP_EXPORT decltype(name)`

But, as I'm not sure if that is an universal fix, someone should confirm that that's the proper way to do it.

**Comment by Berserker66** [comment]
There is another error, the cppolyline.c  and .h are not included in the msvc build, so they get ignored and don't throw errors, there is however a Notch struct that gets initialized wrong for msvc. After fixing that, I got it to fully compile for use in pymunk, or so I thought. It runs through 143 of the 182 pymunk tests before segfaulting. But that could be on the wrapper side.

**Comment by slembcke** [comment]
I think all of these issues should be resolved with the 7.0.3 release.
